### DISCUSSION TITLE
Not allowed to produce messages from a producer hosted in a different namespace to kafka brokers with SCRAM-SHA512 / SASL : SSL Handshake failed

### DISCUSSION THREAD

**Discussion by oesebus** [strimzi/strimzi-kafka-operator#9090] [Q&A] [answered]
Hello 
I am blocked since a couple of days because i am not able to connect to Kafka brokers from a .NET Client using Confluent.Kafka library and MassTransit . 

**Kafka.Sepc :** 

```go
status:
  clusterId: HrBMOd-dSFWlvlJbj8oqHA
  conditions:
    - lastTransitionTime: '2023-09-07T12:51:07.810076146Z'
      status: 'True'
      type: Ready
  listeners:
    - addresses:
        - host: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc
          port: 9092
      bootstrapServers: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092
      name: plain
      type: plain
    - addresses:
        - host: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc
          port: 9093
      bootstrapServers: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9093
      certificates:
        - |
          -----BEGIN CERTIFICATE-----
          MIIFLTCCAxWgAwIBAgIUEDvqBtv7NUWw99VwFZPNiUTg4sswDQYJKoZIhvcNAQEN
          BQAwLTETMBEGA1UECgwKaW8uc3RyaW16aTEWMBQGA1UEAwwNY2x1c3Rlci1jYSB2
          LnN0cmltemkxFjAUBgNVBAMMDWNsdXN0ZXItY2EgdjAwggIiMA0GCSqGSIb3DQEB
          AQUAA4ICDwAwggIKAoICAQDFOLN25c+9ZsyWtKu0XdBxJWRR9BxIoVOLj2XARy8X
          NNqoGkIOH7Y7nC5sjKXZmKBI3ouCDlwNCjgLCufQ/PgcgWnBFzP6jO/oEC6ra20J
          +RR5Lzropc7SrGYhlg5toJFTicSHuO8QOIcfWEWP8jtysobj015tm839ADgcBmCs
          FdiwdechbaatST95ncW8uZHh3L8Hc97dyT4Ds3WjUY7iORqwRXY9z+YwQYMOIrZJ
          2FEGehxYX0kkgEEKfTIbP08FZnUBIplwYT6QnCJteOWPzcMZrZCV5Jn+Vy7Bi62A
          +bu821omS088lp5ZmH9n1yej5dg3i/XxigWd0SfHfKezeDnWciT7T1Ay0y59P75M
          2ZSUP9Xz4jK2UUzpTf3eDgNHetzLTB8XOl6loJcmHz3qjr+TjkSpOMOxyKHNu19q
          TuWEdC6FmRwCjyXU3pjHwKtH9ZoC4dbYl8JZ/mj5q7z8hu7+ETqMH/xKQI5NqRZh
          edcwGTEvfpykJcVyDCeT0OEiYbmAYv5Et5zI6yuZt68IzT7qxs/2pR9BcOW3pIPo
          Zs2VNSmeCe/ChgZLM8mhqip+aBuWDhyO1ON/3YuKckCe/zulpHdtjELOIOkSPWXH
          VQIDAQABo0UwQzAdBgNVHQ4EFgQUB7y4CzC7QsraewgXXJVP3a1ephwwEgYDVR0T
          AQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAQYwDQYJKoZIhvcNAQENBQADggIB
          AHok/JdmOePOtMLd7lZ9H8gm5WUKA+UF+F1kfk7ErKH/1IpWR/H1JUQ5Zzzjoz/Q
          sS6praY6wNf9fBo0BEGLXWsiqUGJSixb7mdVACbTF64Wy0v+SOfVvQqyYLj1MqAh
          33+Q71CAwJvc6nihZKtFzDyUsvkeIQ2SylSrL4OwsiHzNYM44tyLBcuyyZ7NHe7V
          JCvn+wYWE7ZPHpuiIcWoe8bLCPt9jMGu+2FBRjjSnoeHaQ1t+nL6asX62PtXpI0u
          FflNjKFNg94qLUddz1H3DbYYDxsk5O6xvw7N3DnKYuJlcx5SnallRTp3rOPqpCFu
          aqUpccIgHKp58HFka9Ava8EUXP8QKQ57V25gM7PQaB9mBMeSS1dBVIH1C7laFwUJ
          2I164A8YOBsov9xEjGhlo/S48QxS35fs43F8vD+7hFExG3XxGQ2+kqyWbFsOR/SG
          uejj1z2SyO3ucRH19LKZwGhFauaA3J45uRCu+7vilRb19CTrs5NkNy7cmYhB/60K
          rioa2nJ0ux/VrpSfjEV5Tz06AKC773dSE4pTVYDKKfSYkoy9fpbCF2RTnB0XYxtm
          XOhWpQxELRe5bLeAweMtZtrX6imp+KQHEGkv4G6iYRNLbhkdKWLCtveikJCQrN9j
          O50135ZpoHaw1AaODJxykix/2d7UoynKRpyF5uONwlmn
          -----END CERTIFICATE-----
      name: tls
      type: tls
    - addresses:
        - host: kafka-0
          port: 30758
      bootstrapServers: kafka-0:30758
      name: external
      type: external
  observedGeneration: 30
spec:
  entityOperator:
    topicOperator: {}
    userOperator: {}
  kafka:
    authorization:
      superUsers:
        - gambit
      type: simple
    config:
      allow.everyone.if.no.acl.found: true
      auto.create.topics.enable: false
      default.replication.factor: 1
      inter.broker.protocol.version: '3.4'
      log.retention.bytes: 1073741824
      log.retention.hours: 168
      min.insync.replicas: 1
      num.partitions: 1
      offsets.topic.replication.factor: 1
      transaction.state.log.min.isr: 1
      transaction.state.log.replication.factor: 1
      unclean.leader.election.enable: false
    listeners:
      - authentication:
          type: scram-sha-512
        name: plain
        port: 9092
        tls: false
        type: internal
      - name: tls
        port: 9093
        tls: true
        type: internal
      - configuration:
          brokers:
            - advertisedHost: kafka-0
              broker: 0
        name: external
        port: 9094
        tls: false
        type: nodeport
    replicas: 1
    resources:
      limits:
        cpu: 2
        memory: 4Gi
      requests:
        cpu: 500m
        memory: 2Gi
    storage:
      deleteClaim: false
      size: 10Gi
      type: persistent-claim
    version: 3.4.0
  zookeeper:
    replicas: 1
    storage:
      deleteClaim: false
      size: 10Gi
      type: persistent-claim
``` 

```go
spec:
  authentication:
    type: scram-sha-512
  authorization:
    acls:
      - host: '*'
        operations:
          - Describe
          - Read
        resource:
          name: '*'
          patternType: literal
          type: topic
      - host: '*'
        operations:
          - Read
        resource:
          name: '*'
          patternType: literal
          type: group
      - host: '*'
        operations:
          - Create
          - Describe
          - Write
          - Read
        resource:
          name: '*'
          patternType: literal
          type: topic
    type: simple

``` 

This is a singleNode cluster . 
My client service is deployed on the same Kubernetes cluster **but not in the same namespace** 
MyService namespace  = tools
Kafka namespace = kafka

What would be the bootstrapserver URL to provide in such context ?

The operator created some services based on my configuration but i tried several options without any sucess.

**client producer configuration**
```c#
                                  saslConfigurator.SecurityProtocol = SecurityProtocol.SaslPlaintext;
                                    saslConfigurator.Mechanism = SaslMechanism.ScramSha512;
                                    saslConfigurator.Username = configuration.Username;  //testUser
                                    saslConfigurator.Password = configuration.Password;`   // password decoded 
``` 
**Kafka logs error** 

> %3|1694092255.194|FAIL|rdkafka#producer-3| [thrd:sasl_ssl://kafka-0:9094/bootstrap]: sasl_ssl://kafka-0:9094/bootstrap: Failed to resolve 'kafka-0:9094': Name does not resolve (after 5ms in state CONNECT, 16 identical error(s) suppressed)

I ve already tested by targetting  gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092 as bootstrapServer (client side) but not working 

Thanks for your precious help or feedback

**Comment by scholzj** [comment, marked as answer]
You should check and fix the formatting of the information you shared because it is not readable.

**Reply by oesebus** [reply]
Hi , 
Thanks for your quick reaction . is it more clear ? (i removed the code zone)

**Reply by scholzj** [reply]
Not really. It needs to be formatted to show the whitespaces which play a key role in YAML.

**Reply by oesebus** [reply]
Ok done.

**Reply by scholzj** [reply]
Ok, now I can read it.

So, where does your client try to connect? What are you actually trying to do? Your Kafka cluster has 3 listeners:
* One with SCRAM-SHA authentication without TLS encryption
* Second with TLS encryption without authentication
* Third without TLS encryption and without authentication

And then:
* The client configuration you shared seems to enable SCRAM-SHA authentication but not TLS encryption -> that would work if the client connects to `gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092`
* But you say that you get an SSL handshake error. So that suggests you connect to some listener with enabled TLS. But that is not configured in the client and you don't have any listener with TLS encryption and SCRAM-SHA authentication.
* Then your error message has completely different error from connecting to `sasl_ssl://kafka-0:9094` => but this is wrong:
    * This does not match the client configuration you shared 
    * The listener on port 9094 is node port listener, so you should be connecting to `kafka-0:30758`. But it is not really clear where did you get the `kafka-0` advertised hostname. Is it something you configured on some infrastructure level? Because if not, you should not set it.
    * The external listener does anyway not enable TLS encryption nor any authentication. So regardless the address and port you use, it would not work.

So I think you need to first think a bit more about what do you want to do. And then go through all your configuration to doublecheck they are correct but also in sync with each other.

**Reply by oesebus** [reply]
@scholzj 
I fully agree with the setup you described above regarding the diffents listeners and their protocols. 

> The client configuration you shared seems to enable SCRAM-SHA authentication but not TLS encryption -> that would work if the client connects to gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092

For instance , i did a test with gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092 and here is the following error  :
[thrd:sasl_ssl://gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kaf]: sasl_ssl://gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092/bootstrap: SSL handshake failed: error:0A000126:SSL routines::unexpected eof while reading (after 110ms in state SSL_HANDSHAKE, 30 identical error(s) suppressed)

Thanks

**Reply by scholzj** [reply]
But you do not have TLs encryption enabled on port 9092. So you should not enable TLS in your client.

**Reply by oesebus** [reply]
I agree but i didn't enabled TLS in my client . 
SecurityProtocol > SaslPlaintext
Mechanism > SCRAM-SHA512
Regards

**Reply by oesebus** [reply]
At Kafka level , i have this error as well > org.apache.kafka.common.network.InvalidReceiveException: Invalid receive (size = 369295617 larger than 104857600)

**Reply by oesebus** [reply]
> So I think you need to first think a bit more about what do you want to do. And then go through all your configuration to doublecheck they are correct but also in sync with each other.

**The idea is to provide 3 ways to communicate with the brokers :**
- No authentication , no encryption on port 9092
- Authentication ,  encryption on port 9093
- Authentication , encryption  on port 9094

**Ive updated the kafka spec as pasted below** 
```go
listeners:
      - name: plain
        port: 9092
        tls: false
        type: internal
      - authentication:
          type: scram-sha-512 
        name: tls
        port: 9093
        tls: true
        type: internal
      - authentication:
          type: scram-sha-512
        name: external
        port: 9094
        tls: true
        type: nodeport
```

**Reply by scholzj** [reply]
> I agree but i didn't enabled TLS in my client .

Not in the configuration. But the `sasl_ssl` in the URL suggests it is enabled. So it seems to be enabled when you get the error.

> At Kafka level , i have this error as well > org.apache.kafka.common.network.InvalidReceiveException: Invalid receive (size = 369295617 larger than 104857600)

This is the typical error when one side uses TLS and other does not. They don't understand each other and it results in this kind of error.

> The idea is to provide 3 ways to communicate with the brokers

That is of course not completely secure with the insecure listener on port 9092. But technically this is fine. You just need to make sure the clients are configured accordingly.

**Reply by oesebus** [reply]
Hi @scholzj Thanks a lot for giving me the direction . I was able to connect now . My client settings were overriden by an internal libray used by the service and the SASL mechanism was forced at the end . 

But now i jumped into the next issue > Broker: Topic authorization failed 

But the weird thing is that i don't have authorization enabled as far as i know how the ACL's mechanism works : 

```go
spec:
  entityOperator:
    topicOperator: {}
    userOperator: {}
  kafka:
    config:
      allow.everyone.if.no.acl.found: true
      auto.create.topics.enable: false
      default.replication.factor: 1
      inter.broker.protocol.version: '3.4'
      log.retention.bytes: 1073741824
      log.retention.hours: 168
      min.insync.replicas: 1
      num.partitions: 1
      offsets.topic.replication.factor: 1
      transaction.state.log.min.isr: 1
      transaction.state.log.replication.factor: 1
      unclean.leader.election.enable: false
    listeners:
      - name: plain
        port: 9092
        tls: false
        type: internal
      - name: tls
        port: 9093
        tls: true
        type: internal
      - authentication:
          type: scram-sha-512
        name: external
        port: 9094
        tls: false
        type: nodeport
    replicas: 1
    resources:
      limits:
        cpu: 2
        memory: 4Gi
      requests:
        cpu: 500m
        memory: 2Gi
    storage:
      deleteClaim: false
      size: 10Gi
      type: persistent-claim
    version: 3.4.0
  zookeeper:
    replicas: 1
    storage:
      deleteClaim: false
      size: 10Gi
      type: persistent-claim
``` 
Kr

**Reply by scholzj** [reply]
Well, you had authorization enabled in the previous YAMLs. So maybe you just didn't properly update the cluster to disable it? You can do `kubectl get kafka -o yaml` and share the full output to check it.

**Reply by oesebus** [reply]
Here we go : 

```go
kind: Kafka
metadata:
  annotations:
    meta.helm.sh/release-name: kafka
    meta.helm.sh/release-namespace: techno-tools-kafka
  creationTimestamp: "2023-08-28T07:38:34Z"
  generation: 38
  labels:
    app.kubernetes.io/managed-by: Helm
    k8slens-edit-resource-version: v1beta2
  name: gwp-strimzi-cluster
  namespace: techno-tools-kafka
  resourceVersion: "89253875"
  uid: 0658daa7-a7a7-4158-84ab-1b52af49a863
spec:
  entityOperator:
    topicOperator: {}
    userOperator: {}
  kafka:
    config:
      auto.create.topics.enable: false
      default.replication.factor: 1
      inter.broker.protocol.version: "3.4"
      log.retention.bytes: 1073741824
      log.retention.hours: 168
      min.insync.replicas: 1
      num.partitions: 1
      offsets.topic.replication.factor: 1
      transaction.state.log.min.isr: 1
      transaction.state.log.replication.factor: 1
      unclean.leader.election.enable: false
    listeners:
    - name: plain
      port: 9092
      tls: false
      type: internal
    - name: tls
      port: 9093
      tls: true
      type: internal
    - authentication:
        type: scram-sha-512
      name: external
      port: 9094
      tls: false
      type: nodeport
    replicas: 1
    resources:
      limits:
        cpu: 2
        memory: 4Gi
      requests:
        cpu: 500m
        memory: 2Gi
    storage:
      deleteClaim: false
      size: 10Gi
      type: persistent-claim
    version: 3.4.0
  zookeeper:
    replicas: 1
    storage:
      deleteClaim: false
      size: 10Gi
      type: persistent-claim
status:
  clusterId: HrBMOd-dSFWlvlJbj8oqHA
  conditions:
  - lastTransitionTime: "2023-09-08T09:07:26.527457851Z"
    status: "True"
    type: Ready
  listeners:
  - addresses:
    - host: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc
      port: 9092
    bootstrapServers: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9092
    name: plain
    type: plain
  - addresses:
    - host: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc
      port: 9093
    bootstrapServers: gwp-strimzi-cluster-kafka-bootstrap.techno-tools-kafka.svc:9093
    certificates:
    - |
      -----BEGIN CERTIFICATE-----
      MIIFLTCCAxWgAwIBAgIUEDvqBtv7NUWw99VwFZPNiUTg4sswDQYJKoZIhvcNAQEN
      BQAwLTETMBEGA1UECgwKaW8uc3RyaW16aTEWMBQGA1UEAwwNY2x1c3Rlci1jYSB2
      MDAeFw0yMzA4MjgwNzM5MDFaFw0yNDA4MjcwNzM5MDFaMC0xEzARBgNVBAoMCmlv
      LnN0cmltemkxFjAUBgNVBAMMDWNsdXN0ZXItY2EgdjAwggIiMA0GCSqGSIb3DQEB
      AQUAA4ICDwAwggIKAoICAQDFOLN25c+9ZsyWtKu0XdBxJWRR9BxIoVOLj2XARy8X
      NNqoGkIOH7Y7nC5sjKXZmKBI3ouCDlwNCjgLCufQ/PgcgWnBFzP6jO/oEC6ra20J
      +RR5Lzropc7SrGYhlg5toJFTicSHuO8QOIcfWEWP8jtysobj015tm839ADgcBmCs
      FdiwdechbaatST95ncW8uZHh3L8Hc97dyT4Ds3WjUY7iORqwRXY9z+YwQYMOIrZJ
      2FEGehxYX0kkgEEKfTIbP08FZnUBIplwYT6QnCJteOWPzcMZrZCV5Jn+Vy7Bi62A
      +bu821omS088lp5ZmH9n1yej5dg3i/XxigWd0SfHfKezeDnWciT7T1Ay0y59P75M
      r5A6bGuNLT/PzjJ2AWkimGe72cWX74xFtgKVSHZHTcMuBi6EcpNWEpJEWicvecnx
      2ZSUP9Xz4jK2UUzpTf3eDgNHetzLTB8XOl6loJcmHz3qjr+TjkSpOMOxyKHNu19q
      TuWEdC6FmRwCjyXU3pjHwKtH9ZoC4dbYl8JZ/mj5q7z8hu7+ETqMH/xKQI5NqRZh
      edcwGTEvfpykJcVyDCeT0OEiYbmAYv5Et5zI6yuZt68IzT7qxs/2pR9BcOW3pIPo
      Zs2VNSmeCe/ChgZLM8mhqip+aBuWDhyO1ON/3YuKckCe/zulpHdtjELOIOkSPWXH
      VQIDAQABo0UwQzAdBgNVHQ4EFgQUB7y4CzC7QsraewgXXJVP3a1ephwwEgYDVR0T
      AQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAQYwDQYJKoZIhvcNAQENBQADggIB
      AHok/JdmOePOtMLd7lZ9H8gm5WUKA+UF+F1kfk7ErKH/1IpWR/H1JUQ5Zzzjoz/Q
      sS6praY6wNf9fBo0BEGLXWsiqUGJSixb7mdVACbTF64Wy0v+SOfVvQqyYLj1MqAh
      33+Q71CAwJvc6nihZKtFzDyUsvkeIQ2SylSrL4OwsiHzNYM44tyLBcuyyZ7NHe7V
      JCvn+wYWE7ZPHpuiIcWoe8bLCPt9jMGu+2FBRjjSnoeHaQ1t+nL6asX62PtXpI0u
      2I164A8YOBsov9xEjGhlo/S48QxS35fs43F8vD+7hFExG3XxGQ2+kqyWbFsOR/SG
      uejj1z2SyO3ucRH19LKZwGhFauaA3J45uRCu+7vilRb19CTrs5NkNy7cmYhB/60K
      rioa2nJ0ux/VrpSfjEV5Tz06AKC773dSE4pTVYDKKfSYkoy9fpbCF2RTnB0XYxtm
      XOhWpQxELRe5bLeAweMtZtrX6imp+KQHEGkv4G6iYRNLbhkdKWLCtveikJCQrN9j
      O50135ZpoHaw1AaODJxykix/2d7UoynKRpyF5uONwlmn
      -----END CERTIFICATE-----
    name: tls
    type: tls
  - addresses:
    - host: 10.224.0.5
      port: 30758
    bootstrapServers: 10.224.0.5:30758
    name: external
    type: external
  observedGeneration: 38
```

**Reply by scholzj** [reply]
That looks good. You should not get any authorization errors.

**Reply by oesebus** [reply]
At broker level , this is the error i see > 

> User:ANONYMOUS is Denied Operation = Describe from host = 10.244.1.192 on resource = Topic:LITERAL ...

Maybe i have to create a kafkaUser for ANONYMOUS ?
Kr

**Reply by scholzj** [reply]
ANONYMOUS means you do not have authentication enabled on the listener. But without authorization set, you should not get any error. I guess you are likely connecting to some other cluster? Would that be possible?

**Reply by oesebus** [reply]
The client is deployed in the same cluster but not in same namespace .

**Reply by scholzj** [reply]
It is more about whether it connects to the right Kafka cluster or whether you got the YAML of the right Kafka cluster. Because you cannot get an authorization error without authorization enabled.

**Reply by oesebus** [reply]
Finally , i just restarted the kafka broker and then it works :-)
Thanks a lot @scholzj for your explanations and patience ^^. Really appreciate it
Have a nice week end

**Reply by scholzj** [reply]
Great, glad it works. Have a nice weekend.
